### ISSUE TITLE
Don't assume enums are exhaustive

### ISSUE THREAD

**Issue by matklad** [gluon-lang/lsp-types#213]
We currently have code like this:

```rust
#[derive(Debug, Eq, PartialEq, Clone, Deserialize_repr, Serialize_repr)]
#[repr(u8)]
pub enum SignatureHelpTriggerKind {
    Invoked = 1,
    TriggerCharacter = 2,
    ContentChange = 3,
}
```

the problem is that if the client implements a protocol extension and send `4`, this will completely prevent deserialization, instead of graceful degradation. We should do the following instead

```rust
#[serde(transparent)]
pub struct SignatureHelpTriggerKind(u8);
impl SignatureHelpTriggerKind {
    pub const Invoked = SignatureHelpTriggerKind(1);
    pub cosnt TriggerCharacter = SignatureHelpTriggerKind(2);
    pub cosnt ContentChange = SignatureHelpTriggerKind(3);
}
```

We already do this for some string-based enums: 

https://github.com/gluon-lang/lsp-types/blob/7801115725943a002ae271a495123318e88b4c19/src/semantic_tokens.rs#L18

**Comment by lnicola** [comment]
That sounds like a good plan.

**Comment by matklad** [comment]
Short term, I think doing refactor like the one in https://github.com/gluon-lang/lsp-types/issues/213#issue-989190158 is the way to go. 

Long-term, *someone* has to write `lsp.d.ts` file along the lines of [`vscode.d.ts`](https://github.com/microsoft/vscode/blob/main/src/vs/vscode.d.ts), so that we can just generate the whole thing instead of manually duplicating the info from markdown files.

I think it needs: 

* manual or semi-automated "parsing" of LSP markdown spec into `lsp.d.ts` file
* writing some TypeScript code to parse that `d.ts` and to emit some language-independent JSON (more or less serializing TS AST into JSON)
* writing some Rust code to convert from JSON to serde structs

Ideally, this should be done in the protocel iteslf, but there was very little motion on this upstream: https://github.com/microsoft/language-server-protocol/issues/67.

So, organizationally, I think it makes most sense for us to do all the work ourselves, and than just ask upstream if they are willing to pull.

**Comment by lnicola** [comment]
@archseer not really. First of all, it would prevent users from defining protocol extensions, because you wouldn't be able to make new enum variants. Second, I don't know if `serde` knows can deal with those. What might have worked was something like:

```rust
#[repr(u8)]
enum Foo {
    Bar = 1,
    Baz = 2,
    #[serde(other)]
    Extension(u8) = 3,
}
```

but I don't think `#[serde(other)]` works like that, and custom discriminants on tuples isn't stable anyway.
